wgpu-core/src/hub.rs

```diff
@@ -121,32 +129,54 @@ impl<T, I: TypedId> ops::IndexMut<I> for Storage<T, I> {
 impl<T, I: TypedId> Storage<T, I> {
     pub fn contains(&self, id: I) -> bool {
         let (index, epoch, _) = id.unzip();
-        match self.map.get(index as usize) {
-            Some(&(_, storage_epoch)) => epoch == storage_epoch,
-            None => false,
+        match self.map[index as usize] {
+            Element::Occupied(_, storage_epoch) => epoch == storage_epoch,
+            Element::Vacant => false,
+            _ => unreachable!(),
         }
     }
 
-    pub fn insert(&mut self, id: I, value: T) -> Option<T> {
+    pub fn insert(&mut self, id: I, value: T) -> Element<T> {
         let (index, epoch, _) = id.unzip();
-        let old = self.map.insert(index as usize, (value, epoch));
-        old.map(|(v, _storage_epoch)| v)
+        let len = self.map.len();
+        if len <= index as usize {
+            self.map
+                .resize_with(index as usize - len + 1, || Element::Vacant);
+        }
+        std::mem::replace(
+            &mut self.map[index as usize],
+            Element::Occupied(value, epoch),
+        )
     }
 
     pub fn remove(&mut self, id: I) -> Option<T> {
         let (index, epoch, _) = id.unzip();
-        self.map
-            .remove(index as usize)
-            .map(|(value, storage_epoch)| {
-                assert_eq!(epoch, storage_epoch);
-                value
-            })
+        let element = if index as usize == self.map.len() - 1 {
+            self.map.pop().unwrap()
+        } else {
+            std::mem::replace(&mut self.map[index as usize], Element::Vacant)
+        };
+        if let Element::Occupied(value, storage_epoch) = element {
+            assert_eq!(epoch, storage_epoch);
+            Some(value)
+        } else {
+            None
+        }
     }
 
     pub fn iter(&self, backend: Backend) -> impl Iterator<Item = (I, &T)> {
-        self.map.iter().map(move |(index, (value, storage_epoch))| {
-            (I::zip(index as Index, *storage_epoch, backend), value)
-        })
+        self.map
+            .iter()
+            .enumerate()
+            .filter_map(move |(index, x)| {
+                let w = if let Element::Occupied(ref value, ref storage_epoch) = x {
```

Comment:
nit: let's do `match *x` here instead. The construct you have here has implicit borrowing